design/lsm.md

```diff
@@ -0,0 +1,50 @@
+# Log-structured merge-tree (LSM) index
+
+Authored: Sept 21, 2023
+Author: Thor Hansen (@thorfour)
+
+## Abstract
+
+FrostDB was built using a BTree as it's index. This decision was made based on early assumptions and before the query engine
+had taken shape. Now with a better understanding of how FrostDB works in practice it is my recommendation that we replace the Btree index with
+an LSM index.
+
+## Problem
+
+Insertion spends the largest portion of CPU cycles on sorting the individual rows of records into their corresponding granules. However due to the multi-key primary key 
+schemas coupled with dynamic columns, the query engine does not leverage the sorted nature of rows during query time. The scan layer simply passes the records into the query engine.
+Which means the only benefit from sorting the rows is that during compaction a granule already contains the rows that will likely offer the best compaction benefits.
+
+However this means that we're slowing down insertion times for potential gains only during compaction time. We've witnessed things like replay and recovery become dominated by the 
+insert path.
+
+## Proposal
+
+Instead of using a Btree where we sort the individual rows of each record into granules, we instead use an LSM tree where the entire record is simply appended to the first level
+of the LSM index. Because of the immutable nature of the data in FrostDB we do not require all the features of an LSM trree. But are instead lifting the ideas of having different
+levels of compacted data. With the first level being a raw list of arrow records that have been written, and subsequent levels being compacted records.
+
+
+|---------|    |---------|    |---------|
+|         |    |         |    |         |
+|  Record |--->|  Record |--->|  Record |   L0: Raw written Arrow records
+|         |    |         |    |         |
+|_________|    |_________|    |_________|
+
+|---------|    |---------|    |---------|
+|         |    |         |    |         |
+| Parquet |--->| Parquet |--->| Parquet |   L1: Compacted L0 in Parquet files.
+|         |    |         |    |         |
+|_________|    |_________|    |_________|
+
+                Fig 1.
+
+
+### Compaction
+
+Compaction would work much like it does today except instead of compacting at the Granule level it takes all of L0 and compacts those records into a Parquet file (potentially in the future another larger arrow record).
```

Comment:
Again, kind of a moot point since we shove all records/row groups into the query engine, but the big difference here is that the memtable/l0 will be much larger than what we're currently used to in a single compaction given it will be global. How will this affect memory usage?

Additionally, this compaction will also need to merge memtable/l0 with any globally overlapping l1 parts.

Reply:
I don't think the memotables is much larger it would just be a compaction setting that we can tune to adjust our tradeoffs of memory vs CPU much like compaction today

Reply:
Also this compaction won't merge across level boundaries.

Reply:
> I don't think the memotables is much larger it would just be a compaction setting that we can tune to adjust our tradeoffs of memory vs CPU much like compaction today

But currently we only look at the l0 parts of a single granule for a given compaction. The memtable will be equivalent to compacting all l0 parts across all granules at the same time. I agree this would be something to tune, but I think it's important to keep in mind how compactions change.

> Also this compaction won't merge across level boundaries.

OK. Maybe this is a good thing, but again we'll need to change this if we ever move towards more of a strict LSM implementation since with this you're forced to do a linear scan over the whole level. Again, we already do this so I don't think it's an issue. This might be nice to also reduce compaction memory usage. Also if we can guarantee l0 won't overlap by maybe throwing away dynamic columns as sorting columns that'll be even better.

Reply:
Agreed. Definitely something to revisit in the future should we change the querying